warehouse/internal/loadfiles/loadfiles.go

```diff
@@ -572,3 +587,105 @@ func toLoadFile(output LoadFileUpload, job *model.UploadJob) model.LoadFile {
+// and then applies size constraints within each group. The maxSizeMB parameter controls the maximum size of any table within a group.
+func (lf *LoadFileGenerator) GroupStagingFiles(files []*model.StagingFile, maxSizeMB int) [][]*model.StagingFile {
+	// The fields in this struct are determined by the fields being used in the generateBaseRequest
+	type stagingFileGroupKey struct {
+		UseRudderStorage      bool
+		DestinationRevisionID string
+		TimeWindow            time.Time
+	}
+	groups := make(map[stagingFileGroupKey][]*fileWithMaxSize)
+
+	for _, file := range files {
+		key := stagingFileGroupKey{
+			UseRudderStorage:      file.UseRudderStorage,
+			DestinationRevisionID: file.DestinationRevisionID,
+			TimeWindow:            file.TimeWindow,
+		}
+		maxSize := int64(0)
+		for _, size := range file.BytesPerTable {
+			if size > maxSize {
+				maxSize = size
+			}
+		}
+		groups[key] = append(groups[key], &fileWithMaxSize{
+			file:    file,
+			maxSize: maxSize,
+		})
+	}
+
+	result := make([][]*model.StagingFile, 0, len(groups))
+
+	// For each group, apply size constraints
+	for _, group := range groups {
+		result = append(result, lf.groupBySize(group, maxSizeMB)...)
+	}
+	return result
+}
+
+// splits a group of staging files based on size constraints
+func (lf *LoadFileGenerator) groupBySize(files []*fileWithMaxSize, maxSizeMB int) [][]*model.StagingFile {
+	maxSizeBytes := int64(maxSizeMB) * 1024 * 1024 // Convert MB to bytes
+	// Sort by the largest table size in each file
+	slices.SortFunc(files, func(a, b *fileWithMaxSize) int {
+		// Not going with b.maxSize - a.maxSize to avoid int overflow
+		if b.maxSize > a.maxSize {
+			return 1
+		} else if b.maxSize < a.maxSize {
+			return -1
+		}
+		return 0
+	})
+
+	var result [][]*model.StagingFile
+	for len(files) > 0 {
+		// Start a new batch
+		var currentBatch []*model.StagingFile
+		batchTableSizes := make(map[string]int64)
+
+		// Try to add files to the current batch
+		i := 0
+		for i < len(files) {
+			// Check if adding this file would exceed size limit for any table
+			canAdd := true
+			for tableName, size := range files[i].file.BytesPerTable {
+				newSize := batchTableSizes[tableName] + size
+				if newSize > maxSizeBytes {
+					canAdd = false
+					break
+				}
+			}
+
+			if canAdd {
+				// Add file to batch and update table sizes
+				currentBatch = append(currentBatch, files[i].file)
+				for tableName, size := range files[i].file.BytesPerTable {
+					batchTableSizes[tableName] += size
+				}
+				// Remove the file from remaining by moving the last element
+				files[i] = files[len(files)-1]
+				files = files[:len(files)-1]
+			} else {
+				i++
+			}
+		}
+
+		// If we couldn't add any files to the batch, add the first file anyway
+		// This ensures we make progress even with files larger than maxSizeBytes
+		if len(currentBatch) == 0 {
+			currentBatch = append(currentBatch, files[0].file)
+			files = files[1:]
+		}
```

Comment:
Wondering if this logic can be included in iteration of `for i < len(files) {` itself. Imagine all files maxSize exceeding configured maxSize. Then we are looping for all and then taking out 1 here. And then again we loop through all files. This is worst case scenario of n^2.

Reply:
Moved the logic inside the for loop

Reply:
Thanks. My bad, it was not n^2 even with this code, missed reading the break part.